hashicorp/consul
# fix(mesh-gateway): remove deregistered service from mesh gateway

### Description
- Deregistered service persists in the mesh-gateway's clusters. This occurs when the mesh-gateway `handleUpdate` un-watches a service without removing it from the ServiceGroups.

- This PR also fixes another issue of exporting unregistered services (details below). An exported service is still being available to importing service, while it is deregistered and removed from the catalog of exporting cluster.

### Testing & Reproduction steps

* Establish peering between 2 clusters
1. Export a service from cluster A to cluster B ( A ---> B)
2. Verify the service in B can access service in A
3. Deregister the service in A
4. `Export~service` still exists in mesh gateway of A. Further, the service in B can still access the exported service, although it has been deregistered in cluster A.


* Single DC
1. Launch a consul server agent
2. Create a mesh gateway and make sure mesh gateway is healthy in consul UI
3. Register a service and wait till the service is added to the gateway clusters
4. Deregister the service and the service still exists in mesh gateway clusters

### Links

### PR Checklist
